### DISCUSSION TITLE
Timed discharge based on Octopus Outgoing price

### DISCUSSION THREAD

**Discussion by nicwats** [wills106/homeassistant-solax-modbus#108] [General]
Morning, 

Just wondering if anyone has their system working in the same way as I'm trying to get mine set up and has any pointers....
I've got a X1-AC-3.6 inverter, 4.5kw battery and 5kw solar array, currently just with Home Assistant set up with a bunch of monitoring of the system productivity etc. 
I'm signed up to the Octopus Flexible tariff for incoming power which is currently sat at the price cap of 28p/kw as a flat rate. 
I'm also signed up to the Octopus Agile Outgoing tariff which has a half hourly variable rate for the power that I feed back to the grid. In some instances, this is up at over 60p/kw in the evenings. 
What I'm wondering is if through this setup, I can have the battery run as normal most of the time, i.e. fill up from Solar then solar feed the grid, then as the sun goes down, the house starts drawing the battery until it's empty.... But on cloudy days/through winter where the battery barely fills anyway, control the battery to fill for a few hours prior to the evening peak and then discharge when the Outgoing price is at double what I pay for the incoming in order to recoup a bit of money. 
From everything I've read it sounds plausible and shouldn't be too difficult but I can't see any comments from anyone doing similar. 

Thanks,
Nic

**Comment by wills106** [comment]
Simplest way to charge is to change it to "Force Time" mode with a charge window set and a time period set, don't forget to set "select.solax_allow_grid_charge"

I have only just implemented discharging on the Gen3 you will need to be on the Beta 0.6.0 releases and see WiKi for the basics on how to use it.

**Reply by wills106** [reply]
I have started to update the Wiki with instructions on Discharging the Gen3

**Comment by nicwats** [comment]
Cheers! 
Might be a little off-topic but just wondering if it's something you've got any experience in. My old energy supplier in the UK was a company called Social Energy. They used to run a 'grid trading' system whereby there was a box on the wall connected to the meters, and battery inverter. Their principle was fairly simple in that they used to provide a service to the grid as fast frequency response to help mitigate any fluctuations in grid power frequency by charging or discharging the battery and in doing so, provided a share of the profits as well as the standard feed in tariff. 
They went under with the energy crisis and claim to still be doing the fast frequency response but it's not something I'm signed up to any more. They have left the box on the wall and so I've started tearing it down a bit to see how they controlled the battery and if there is a better way. In the box, there's a standard SDM230 modbus meter, a freqency reader and then a beaglebone board which obviously did all of the communication and interpretation. 
I'm just working on getting onto the board to look at their code as I'm wondering if they use the same principle of read/write of the inverter or whether they have a different way of controlling the charge and discharge. 
I am far from expert in any of this and am no coder but I'm hoping there might be a nice way to pull the modbus data from the multiple meters it was connected to as well as control the battery, all via home assistant. 
Apologies for the essay but if you have any thoughts or pointers then it'd be great!

**Reply by wills106** [reply]
Without seeing how it's wired it's hard to say.

Not sure why they have a SDM230 Modbus meter and a frequency reader?
Unless they are trying to do some feedback loop with something else?

**Reply by nicwats** [reply]
I'll snap a picture so you can see it. The frequency reader that's installed I think is far more sensitive than a standard Modbus and allows for much quicker control of the battery which was a requirement of the fast frequency response.

**Reply by nicwats** [reply]
Hopefully it shows up ok. I've been trying to figure out exactly how it works as when it was working properly, it had a decent interface with a lot of good data as well as the ability to control it fairly seamlessly. The top box is actually labelled as a frequency controller rather than just a meter so I'm assuming that it's the beaglebone that controls the charge/discharge by the inverter in much the same way as your current process but just based on frequency response rather than user control or price. 

![IMG_0101](https://user-images.githubusercontent.com/108125210/185412460-256bca1a-c37a-498b-943d-6f81732bae13.jpg)

**Reply by wills106** [reply]
Not really sure what's going on there. The "Frequency Meter" only seems to be connected to the Eastron, Inverter and Beaglebone. Can't see any mains going to it.

I am guessing the grey box next to the Eastron is the 5v PSU for the Beaglebone?

**Reply by nicwats** [reply]
Yeah, I've swapped out the 5v for a 12v and removed the beaglebone and replaced it with the waveshare device so that it's all in a nice neat housing. Connection to the mains is the big black cable on the bottom right with the power cable from the inverter as the chunky white one on the left. 
I assume that this old system must work on very similar principles as the only connection to the inverter is via those same two pins on the COM port. Interesting that the pins for the Eastron and inverter feed are connected to the same points on the frequency meter.

**Comment by nicwats** [comment]
@wills106 Can you just confirm that for the X1-AC-3.6, the best Wiki to follow is the Gen 4 installation instructions? I've got the Waveshare Eth adapter and can connect that to pins 4 & 5 as per the Gen 4 instructions unless that's an issue?

**Reply by wills106** [reply]
Just checked the manual on P16 of pdf (28 - 29 on real copy) and it's 4 & 5. Your Baud rate should be 115200

**Reply by wills106** [reply]
How did you get on?

**Reply by nicwats** [reply]
Frustratingly still trying to get windows set up so I can see it. I’ve also got to check The cables for the existing system as it seems to run back and forth through crossover cables so I want to make sure I’m picking up pins 4&5 correctly. What’s the pdf manual you mention? Is there an installers manual available or something? I’d been trying to find one for a while.

**Reply by wills106** [reply]
It's on the product page on the [SolaX Website](https://www.solaxpower.com/ac-charger/) at the bottom. "X1-AC User Manual"

**Reply by nicwats** [reply]
Thanks, hadn't realised that it went to that level of detail. 
<img width="1058" alt="Screenshot 2022-09-13 at 10 34 44" src="https://user-images.githubusercontent.com/108125210/189867518-cf562539-0e86-474d-b5f1-a2c73dfb7ea2.png">
Think I've got it up and running now but not certain. I've got the integration installed and it's showing 5x devices. That about right?

**Reply by nicwats** [reply]
It doesn't seem to be pulling the information correctly from the sensors yet.

**Reply by wills106** [reply]
If the Integration only has 5 entities do you have the following error in your log "unrecognized inverter type - serial number : {your_serial_number_here}"

If so what's the first 4 digits?

**Reply by nicwats** [reply]
<img width="972" alt="Screenshot 2022-09-13 at 13 12 22" src="https://user-images.githubusercontent.com/108125210/189898070-1096a90e-c007-47d8-9d5b-9e7d21efbc79.png">

**Reply by wills106** [reply]
Seems like the Waveshare isn't talking to the Inveter.

Have you double checked what the Baudrate and address is on the Inverter? As it might have been changed as part of the previous setup?

Is the rest of the Waveshare settings set the same as on the Wiki?

**Reply by nicwats** [reply]
Baudrate is set correctly, do you mean the address on the inverter or the address on the waveshare? And yes, Waveshare settings are the same except I've set it as a fixed IP rather than DHCP. Can't imagine that would matter though?

**Reply by wills106** [reply]
On the Inverter, the default is meant to be 1. But could be set to something else?

**Reply by wills106** [reply]
Did you manage to get any further with your Inverter?

**Reply by nicwats** [reply]
You were right, address was previously set to 4. I've switched it to 1 and it's now communicating properly. Just trying to get my head around controls with the wiki.

**Reply by wills106** [reply]
You should be able to [follow this](https://github.com/wills106/homeassistant-solax-modbus/wiki/Gen2-&-Gen3-Modes-of-operation#grid-export-gen-3-only). You are missing `number.solax_battery_minimum_capacity_grid_tied` but you have the rest of the controls.

I think the second automation example is probably the best overall one.

**Reply by nicwats** [reply]
All seems to be up and running now. Thanks for the support.
One thing that I'm still trying to work out is whether there was a better way of doing this that Social Energy, the hub I mentioned previously, was using. Their beaglebone was only connected via what looked like the ground and RS485A terminals. Given their approach in running rast response for grid charge and discharge, they've either absolutely smashed the 10,000 available EEPROM writes, or there's a different way to control it.

**Reply by wills106** [reply]
They might have been controlling the power meter?

Perhaps to force the X1AC to export they were making the meter appear to be importing from the grid so the Inverter would match that?

**Reply by nicwats** [reply]
Quite a nice simple solution if that's the case. In general, it seemed to sit mainly in a self-use mode to take best advantage of the battery & solar generation but I guess there was also a force-charge period whereby it would fill up before known peaks and then discharge based on a dummy feed of grid import as required.
